Task 4

Neythalion opened the pull request:
added db connection to handlers
added addProduct handler
added json schema validation for addProduct handler

Commit:
Task 4
added db connection to handlers
added addProduct handler
added json schema validation for addProduct handler

Neythalion commented:
[Link on Task 4](https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task4-rds/task.md)

[Link to FE PL](https://github.com/Neythalion/nodejs-aws-fe/pull/3)
[Link on APP](https://d2psr23kjbv8h9.cloudfront.net/)

API:
[Get Products](https://xyxr3kd5v4.execute-api.eu-west-1.amazonaws.com/dev/products)
[Get Product by ID](https://xyxr3kd5v4.execute-api.eu-west-1.amazonaws.com/dev/product/9195882d-a421-4ae8-8992-478aae476cc3)
[Get Product by ID](https://xyxr3kd5v4.execute-api.eu-west-1.amazonaws.com/dev/product/9195882d-a421-4ae8-8992-478aae476cb2) (with 404)
[Add Product](https://xyxr3kd5v4.execute-api.eu-west-1.amazonaws.com/dev/product)

Postgresql db was created
All handlers connected to this db
App receives products from db
Add product handler was implemented with json schema validation middleware
All handlers return 500 on error
All handlers log events into console
Implemented transaction in Add Product hander

VarvaraZadnepriak commented on product-service/serverless.yml:
```diff
@@ -6,6 +6,14 @@ provider:
   region: eu-west-1
 
+  # dbConfig.json with private data was not added to repo
+  environment:
```
[proposal]
It's great that db settings is not commited to the repository, but I still propose to use .env file because it's more common than json

VarvaraZadnepriak commented on product-service/src/handlers/addProduct.js:
```diff
@@ -0,0 +1,58 @@
+};
+
+const addProduct = async event => {
+    console.log(`Lambda invocation - addProduct - product: ${event.body} - event: `, event);
```
[proposal]
Logging params, catching error code repeats across all lambdas => may be a good idea to create helper function for supporting DRY principle

VarvaraZadnepriak commented on product-service/src/handlers/addProduct.js:
```diff
@@ -0,0 +1,58 @@
+import { Client } from 'pg';
+
+const { PG_HOST, PG_PORT, PG_DATABASE, PG_USERNAME, PG_PASSWORD } = process.env;
+const dbConfig = {
```
[proposal]
DRY principle - config the same across all lambdas

VarvaraZadnepriak commented:
Please integrate list and creation form in your FE application: 
https://d2psr23kjbv8h9.cloudfront.net/admin/products
https://d2psr23kjbv8h9.cloudfront.net/admin/product-form/

Commit:
Task 4
moved db settings to .env
added common db settings to constants
fixed add product
added logger middleware
fixed product schema
fixed add product validator
